crates/rustc_codegen_spirv/src/linker/dce.rs

```diff
@@ -37,7 +37,13 @@ fn spread_roots(module: &Module, rooted: &mut FxHashSet<Word>) -> bool {
     for func in &module.functions {
         if rooted.contains(&func.def_id().unwrap()) {
             for inst in func.all_inst_iter() {
-                any |= root(inst, rooted);
+                if !instruction_is_pure(inst) {
+                    any |= root(inst, rooted);
+                } else if let Some(id) = inst.result_id {
+                    if rooted.contains(&id) {
+                        any |= root(inst, rooted);
```

Comment:
hrm, this might be dreadfully inefficient, some profiling might be nice here! (how many times `spread_roots` is called before/after this change). It's probably way more efficient to iterate over instructions backwards instead of forwards.

Reply:
Yep, just checked, when running the `multibuilder` example on `main`, `dce` is called three times, and in each of those, `spread_roots` is called `[3, 4, 1]` times. With this PR, it's `[8, 8, 5]`, a significant cost (definitely not zero-cost, especially with the last one being a no-op). But, reversing the iteration order of the function's instructions, it drops back down to `[3, 4, 1]`.

Reply:
Hm. So do the reversing of iteration and write a comment saying why it is so?